lib/shelley/test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs

```diff
@@ -47,6 +67,155 @@ spec = describe "NetworkLayer regression test #1708" $ do
                     race_ (watchNodeTip nl (putMVar waiter)) (takeMVar waiter)
             void $ waitAnyCancel tasks
 
+    describe "Observer" $ do
+        it "can fetch all observed keys, but not any other keys"
+            $ property $ \keys1 keys2 -> monadicIO $ do
+                (observer, refresh, _trVar) <- run mockObserver
+                run $ mapM_ (startObserving observer) keys1
+                run $ refresh True
+
+                let allNothing = fromKeysWith (const Nothing)
+                let unobservedKeys = Set.difference keys2 keys1
+                unobservedValues <- run $ queryKeys observer unobservedKeys
+
+                observedValues <- run $ queryKeys observer keys1
+
+                assertEqual "observed keys return expected values"
+                    observedValues
+                    (fromKeysWith (Just . length) keys1)
+
+                assertEqual "unobserved keys are all Nothing when queried"
+                    unobservedValues
+                    (allNothing unobservedKeys)
+
+        describe "typical use" $ beforeAll mockObserver $ do
+            -- Using monadic-property tests here /just/ for the sake of testing
+            -- with multiple keys seem worthless.
+            --
+            -- State machine tests might be suitable on the other hand...
+            --
+            -- NOTE: We make sure to test conditions both before and after
+            -- calling @refresh@, as it can be called arbitrarily without our
+            -- (the observers') knowledge.
+            let k = ("k"::String)
+            let v = length k
+            describe "startObserving" $ do
+                it "(query k) returns Nothing before startObserving"
+                    $ \(observer, refresh, trVar) -> do
+                        (query observer k) `shouldReturn` Nothing
+                        trVar `shouldHaveTraced` []
+                        refresh True
+                        (query observer k) `shouldReturn` Nothing
+                        trVar `shouldHaveTraced`
+                            [ MsgWillFetch Set.empty
+                            , MsgDidFetch Map.empty
+                            ]
+
+                it "(query k) returns v after (startObserving k >> refresh)"
+                    $ \(observer, refresh, _) -> do
+                        startObserving observer k
+                        refresh True
+                        let expectedValue = length k
+                        (query observer k)`shouldReturn` Just expectedValue
+
+                it "traced MsgAddedObserver, MsgWillFetch, MsgDidFetch"
+                    $ \(_, _, trVar) -> do
+                        trVar `shouldHaveTraced`
+                            [ MsgAddedObserver k
+                            , MsgWillFetch $ Set.singleton k
+                            , MsgDidFetch $ Map.singleton k v
+                            ]
```

Comment:
I don't get this test, unless this test depends on the previous ones which would be quite bad / misleading. Tests should work in isolation.

Reply:
I wanted to experiment with slightly different usage of `describe`/`it` for more readable test output.

Here we get:
```haskell
        when refresh fails
          (query k) returns the existing v
          only MsgWillFetch is traced
```
which I think is quite neat.

(And in total:)
```
      typical use
        startObserving
          (query k) returns Nothing before startObserving
          (query k) returns v after (startObserving k >> refresh)
          traced MsgAddedObserver, MsgWillFetch, MsgDidFetch
        calling startObserving a second time
          (query k) is still v
        when refresh fails
          (query k) returns the existing v
          only MsgWillFetch is traced
        stopObserving
          makes (query k) return Nothing
```

With no shared state, the whole `typical use` tree would be a single `it`.

I'll admit it gets a bit weird here though, where the `refresh False` belongs to the whole `when refresh fails` tree, but is run in the first `it`. (We cannot have it as a `runIO` or `beforeAll`, because we cannot access the `refresh` function from outside an `it`).

Reply:
We could probably add a `Arbitrary (Observer, refresh)`, and _do_ run most of these checks in isolation though 🤔

Reply:
I admit that the separation is need, but it is puzzling. So at the very least, there should be a clear note as comment explaining basically what you just explained here.